R/parse-body.R

```diff
@@ -16,230 +12,475 @@ postBodyFilter <- function(req){
+#' #* @parser multi
+#' #* @parser rds
+#' pr <- plumber$new()
+#' pr$handle("GET", "/upload", function(rds) {rds}, parsers = c(parser_multi(), parser_rds()))
+#' }
 #' @export
 parser_query <- function() {
-  function(value, content_type = NULL, ...) {
-    charset <- getCharacterSet(content_type)
-    value <- rawToChar(value)
-    Encoding(value) <- charset
-    parseQS(value)
-  }
+  parser_text(parseQS)
 }
 
 
+#' @describeIn parsers JSON parser
+#' @export
+parser_json <- function(...) {
+  parser_text(function(txt_value) {
+    safeFromJSON(txt_value, ...)
+  })
+}
 
 
-#' TEXT
-#' @rdname parsers
+#' @describeIn parsers Helper parser to parse plain text
+#' @param parse_fn function to further decode a text string into an object
 #' @export
-parser_text <- function() {
+parser_text <- function(parse_fn = identity) {
+  stopifnot(is.function(parse_fn))
   function(value, content_type = NULL, ...) {
     charset <- getCharacterSet(content_type)
-    value <- rawToChar(value)
-    Encoding(value) <- charset
-    value
+    txt_value <- rawToChar(value)
+    Encoding(txt_value) <- charset
+    parse_fn(txt_value)
   }
 }
 
 
+#' @describeIn parsers YAML parser
+#' @export
+parser_yaml <- function(...) {
+  parser_text(function(val) {
+    if (!requireNamespace("yaml", quietly = TRUE)) {
+      stop("yaml must be installed for the yaml parser to work")
+    }
+    yaml::yaml.load(val, ..., eval.expr = FALSE)
+  })
+}
 
-
-#" RDS
-#' @rdname parsers
+#' @describeIn parsers Helper parser that writes the binary post body to a file and reads it back again using `read_fn`.
+#'   This parser should be used when reading from a file is required.
+#' @param read_fn function used to read a the content of a file. Ex: [readRDS()]
 #' @export
-parser_rds <- function() {
-  function(value, filename, ...) {
+parser_read_file <- function(read_fn = readLines) {
+  stopifnot(is.function(read_fn))
+  function(value, filename = "", ...) {
     tmp <- tempfile("plumb", fileext = paste0("_", basename(filename)))
-    on.exit(file.remove(tmp), add = TRUE)
+    on.exit({
+      file.remove(tmp)
+    }, add = TRUE)
     writeBin(value, tmp)
-    readRDS(tmp)
+    read_fn(tmp)
   }
 }
 
+#' @describeIn parsers CSV parser
+#' @export
+parser_csv <- function(...) {
+  parser_read_file(function(tmpfile) {
+    if (!requireNamespace("readr", quietly = TRUE)) {
+      stop("`readr` must be installed for `parser_csv` to work")
+    }
+    readr::read_csv(tmpfile, ...)
+  })
+}
 
 
+#' @describeIn parsers TSV parser
+#' @export
+parser_tsv <- function(...) {
+  parser_read_file(function(tmpfile) {
+    if (!requireNamespace("readr", quietly = TRUE)) {
+      stop("`readr` must be installed for `parser_tsv` to work")
+    }
+    readr::read_tsv(tmpfile, ...)
+  })
+}
 
-#" MULTI
-#' @rdname parsers
+
+#' @describeIn parsers RDS parser
+#' @export
+parser_rds <- function(...) {
+  parser_read_file(function(tmpfile) {
+    readRDS(tmpfile, ...)
```

Comment:
Could maybe use `rawConnection` to avoid writing to disk

Reply:
Boo. `readRDS()` does not work with a `rawConnection()`.